benchmarks/key_generator.hpp

```diff
@@ -16,99 +16,209 @@
+   * @param output_begin Start of the output sequence
+   * @param output_end End of the output sequence
+   * @param exec_policy Thrust execution policy this operation will be executed with
+   */
+  template <typename Dist, typename OutputIt, typename ExecPolicy>
+  void generate(Dist dist, OutputIt out_begin, OutputIt out_end, ExecPolicy exec_policy)
+  {
+    using value_type = typename std::iterator_traits<OutputIt>::value_type;
+
+    if constexpr (std::is_same_v<Dist, dist_type::unique>) {
+      thrust::sequence(exec_policy, out_begin, out_end, 0);
+      thrust::shuffle(exec_policy, out_begin, out_end, this->rng_);
+    } else if constexpr (std::is_same_v<Dist, dist_type::uniform>) {
+      size_t num_keys = thrust::distance(out_begin, out_end);
+
+      thrust::counting_iterator<size_t> seeds(this->rng_());
+
+      thrust::transform(exec_policy,
+                        seeds,
+                        seeds + num_keys,
+                        out_begin,
+                        [*this, dist, num_keys] __host__ __device__(size_t const seed) {
+                          RNG rng;
+                          thrust::uniform_int_distribution<value_type> uniform_dist(
+                            0, num_keys / dist.multiplicity);
+                          rng.seed(seed);
+                          return uniform_dist(rng);
+                        });
+    } else if constexpr (std::is_same_v<Dist, dist_type::gaussian>) {
+      size_t num_keys = thrust::distance(out_begin, out_end);
+
+      thrust::counting_iterator<size_t> seq(this->rng_());
+
+      thrust::transform(exec_policy,
+                        seq,
+                        seq + num_keys,
+                        out_begin,
+                        [*this, dist, num_keys] __host__ __device__(size_t const seed) {
+                          RNG rng;
+                          thrust::normal_distribution<> normal_dist(
+                            static_cast<double>(num_keys / 2), num_keys * dist.skew);
+                          rng.seed(seed);
+                          auto val = normal_dist(rng);
+                          while (val < 0 or val >= num_keys) {
+                            // Re-sample if the value is outside the range [0, N)
+                            // This is necessary because the normal distribution is not bounded
+                            // might be a better way to do this, e.g., discard(n)
+                            val = normal_dist(rng);
+                          }
+                          return val;
+                        });
+    } else {
+      // TODO static assert fail
     }
-    case dist_type::UNIFORM: {
-      std::uniform_int_distribution<Key> distribution{1, static_cast<Key>(num_keys / Multiplicity)};
+  }
+
+  /**
+   * @brief Overload of 'generate' which automatically selects a suitable execution policy
+   */
+  template <typename Dist, typename OutputIt>
+  void generate(Dist dist, OutputIt out_begin, OutputIt out_end)
```

Comment:
Does this sync internally or do users have to sync on their own to make sure the data generation is complete? The same question for `dropout`.